### ISSUE TITLE
Improve exception safety with smart pointers

### ISSUE THREAD

**Issue by elfring** [openalpr/openalpr#252]
Would you like to [wrap any pointer data members](https://github.com/openalpr/openalpr/blob/1a64daf8a61793c20d9362a50eaaec2de08d682c/src/openalpr/alpr_impl.cpp#L31 "Update candidate: AlprImpl class") with the template class "[std::unique_ptr](https://en.wikipedia.org/wiki/Smart_pointer#unique_ptr "a smart pointer")"?

**Comment by matthill** [comment]
Does this require C++11?  I'm hesitant to do that, since there's a number of users compile OpenALPR on old compilers

**Comment by elfring** [comment]
How do you think about to [reuse an approach by Howard E. Hinnant](https://howardhinnant.github.io/unique_ptr03.html "Tutorial for unique_ptr")?

**Comment by matthill** [comment]
Not sure, it looks like this requires C++03 or C++11 as well?

**Comment by elfring** [comment]
Do you understand the descriptions by Howard E. Hinnant?

**Comment by matthill** [comment]
Perhaps not.  It seems that the article is explaining the benefits/advantages of unique_ptr over auto_ptr.  Or am I misunderstanding?

**Comment by elfring** [comment]
Can you follow the descriptions why [a class like "std::unique_ptr"](https://isocpp.org/wiki/faq/cpp11-library "Description for useful C++ classes") should be reused more often nowadays?

**Comment by matthill** [comment]
Yes, I believe I understand.  It makes sense.  However, my understanding is that it requires C++11, which I don't want to require.

**Comment by elfring** [comment]
How do you think about to look at similar functionality then from other class libraries if you still need to support software around older versions of the programming language "C++"?

**Comment by matthill** [comment]
I'm not opposed.  If it simplifies/improves the code and is not a burden for people to compile, then it makes sense.  I didn't link in larger libraries like Boost in the past, since it's such a pain for people to get it linked in especially on Windows.

**Comment by elfring** [comment]
Do the chances improve then for the reuse of any more functionality from [the standard "ISO/IEC 14882:2011"](https://en.wikipedia.org/wiki/C%2B%2B11 "Description for C++11")?

**Comment by matthill** [comment]
Markus, I appreciate your expert advice, and the fact that your phrasing each suggestion in the form of a question.  Please, feel free to directly state your opinion about what should be done -- I am honestly having trouble figuring out what you're suggesting and I promise I won't take offense.

**Comment by elfring** [comment]
I am using questions often for my personal communication style. - I leave the corresponding choice for an answer to you as usual.

I am also aware that some software developers do occasionally not like special question series because they might not fit to their current mood. - I am just curious when the time will be ready to reuse popular functionality from more recent C++ support.

**Comment by matthill** [comment]
Ahh, yes, I understand.  I think it's a great communication style :) I figured you were trying to soften the blow of a comment such as "You're doing this completely wrong, here's what you need to do."  Which I actually wouldn't mind at all.

I'm not sure about when exactly we should move to C++11... it just seems like compiler support is still too new: http://en.cppreference.com/w/cpp/compiler_support  and there's a number of folks who need to compile this on old stuff (e.g., a camera with an embedded OS).  It would be a shame to disqualify them from using OpenALPR unless there's a really compelling reason.

**Comment by elfring** [comment]
How do you think about to handle such special cases by corresponding configuration checks in an advanced software build system?

**Comment by matthill** [comment]
I'm definitely not opposed, as long as it is doing something useful and doesn't make the code significantly more complicated.

**Comment by elfring** [comment]
Which software build systems do you reuse so far?

**Comment by elfring** [comment]
>… it just seems like compiler support is still too new …

By the way: Would it make sense to consider cross-compilation by newer build tool versions for older software and hardware environments?

**Comment by matthill** [comment]
Cmake handles the build, I use gcc 4.8.4 to compile.  Some folks use Clang, and of course Visual Studio on Windows.  

Cross-compile could work.  Although, what would probably happen is someone would try to compile it on the target platform and then file an issue when it doesn't work.

**Comment by elfring** [comment]
How do you think about to store a build specification which can be similar to [the script "CMakeLists.txt"](https://github.com/openalpr/openalpr/blob/28e366a16c35cb094ac6343d8865b37e23212dcb/src/CMakeLists.txt#L1 "CMake build script") also in the main project directory?
